### DISCUSSION TITLE
[Bug]: NeuroDeskApp Apple M1 spm12 gui crashes on opening

### DISCUSSION THREAD

**Discussion by scott-love** [neurodesk/neurodesk.github.io-hugo#382] [Bug Report]
### What platform are you seeing the problem on?

Local PC

### What happened

Hi all,
When using Sonoma 14.1.1 ; Apple M1 Max; NeurodeskApp 1.2.1 the SPM12 GUI fails to open or crashes on opening. I do not have this problem with an Intel based Mac.

The screenshot attached shows the last terminal output before the terminal crashes.

<img width="1219" alt="Screenshot 2023-11-17 at 13 43 47" src="https://github.com/NeuroDesk/neurodesk.github.io/assets/12942961/92effb0b-bbf9-46de-ba87-65bab1920950">

**Comment by stebo85** [comment]
Dear @scott-love,

I can reproduce this on my end. It looks like a limitation of the x86 emulation on ARM hardware in combination with Matlab:
<img width="1097" alt="image" src="https://github.com/NeuroDesk/neurodesk.github.io/assets/4021595/bc6e67b2-9385-4ebd-a7de-09720e4551bc">

I currently have no solution idea for this - if anyone has any idea what this could be, let me know.

**Comment by stebo85** [comment]
Dear @scott-love,

Thanks to a few very helpful people at Mathworks, we found out what the problem is and how to fix it: The emulation frameworks do not provide a correct processor frequency in /proc/cpuinfo and that causes the problem reported.

The quick hack around for now is:

create a file called cpuinfo in your Neurodesk homedirectory with the following content:
```
processor	: 0
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 1
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 2
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 3
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 4
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

```

Then add this to the end of your ~/.bashrc file:
```
export SINGULARITY_BINDPATH=$SINGULARITY_BINDPATH,/home/${NB_USER}/cpuinfo:/proc/cpuinfo
```

We are working on a more permanent fix now that doesn't require hardcoding this.

Can you please let us know if this fix works for you for now?

**Comment by scott-love** [comment]
Hi @stebo85,

Thanks for working on this! I implemented the quick fix and it enabled me to open the SPM GUI, which is great. I haven't tested any further. Being new to Neurodesk it took me a second to figure out exactly how to implement it. So I am adding here exactly what I did.

I first opened Neurodesk (v1.2.1) and selected the terminal option/interface.

```
cd jovyan/ 
cat > cpuinfo
processor	: 0
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 1
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 2
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 3
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778

processor	: 4
BogoMIPS	: 48.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint
CPU implementer	: 0x00
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0
cpu MHz         : 2245.778
```

Press CTRL+d to close and save this new cpuinfo file.

```
cat >> .bashrc
echo "export SINGULARITY_BINDPATH=$SINGULARITY_BINDPATH,/home/${NB_USER}/cpuinfo:/proc/cpuinfo"
```

Press CTRL+d to close and save the updated .bashrc file.

I was then able to open a Neurodesktop instance and select Neurodesk/Functional Imaging/spm12/spm12CUI r7771, which successfully opened the SPM GUI.

Something that positively surprised me here is that this fix is persistent. I didn't not need to redo the fix after closing and reopening my local session.

**Reply by stebo85** [reply]
great to hear it worked for you :) yes, the fix will be permanent for you and sorry for not describing it in more detail. Will add this fix im the next version.

**Comment by stebo85** [comment]
Dear @scott-love,

Closing this discussion now :) Neurodesk-app version 1.3.0 and neurodesktop version 2023-11-28 fix this problem now. 

Thank you for reporting it!
Kind regards
Steffen
